pkg/indexerpoc/task.go

```diff
@@ -0,0 +1,449 @@
+	topics := [][]string{}
+	if len(contract.GetTopics()) > 0 {
+		topics = append(topics, contract.GetTopics())
+	}
+
+	filter := NewFilter(addresses, topics)
+
+	task := &task{
+		ctx:         ctx,
+		src:         src,
+		contract:    contract,
+		filter:      filter,
+		db:          db,
+		batchSize:   batchSize,
+		concurrency: 1,
+	}
+
+	// Apply options.
+	for _, opt := range opts {
+		opt(task)
+	}
+
+	// Load existing state or create new state.
+	// TODO: This would be replaced with a block/task tracker.
+	state, err := db.GetState(ctx, contract.GetName(), fmt.Sprintf("%d", contract.GetChainID()))
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			// If no state exists, start from contract-defined block.
+			// If state exist we must force the starting block to prevent missing any data.
+			initialBlockNum := contract.GetStartBlock()
+
+			if initialBlockNum > 0 {
+				initialBlockNum--
+			}
+
+			// Get hash for initial block if not starting from genesis.
+			var initialHash common.Hash
+			if initialBlockNum > 0 {
+				initialHash, err = src.GetBlockHash(ctx, initialBlockNum)
+				if err != nil {
+					return nil, fmt.Errorf("getting hash for initial block on chain %d: %w",
+						src.GetChainID(), err)
+				}
+			}
+
+			task.state = &taskState{
+				ContractName: contract.GetName(),
+				NetworkName:  src.GetNetworkName(),
+				ChainID:      src.GetChainID(),
+				BlockNumber:  initialBlockNum,
+				BlockHash:    initialHash,
+			}
+		} else {
+			return nil, fmt.Errorf("getting state for %s on chain %d: %w",
+				contract.GetName(), src.GetChainID(), err)
+		}
+	} else {
+		task.state = state
+	}
+
+	if state.ChainID != src.GetChainID() {
+		return nil, fmt.Errorf("chain ID mismatch for %s: stored %d, current %d",
+			contract.GetName(), state.ChainID, src.GetChainID())
+	}
+
+	return task, nil
+}
+
+// nextState performs one indexing run, handling blockchain data retrieval and processing.
+// It returns an error if there is nothing new to index, or if there is a reorg.
+func (t *task) nextState() error {
+	ctx := t.ctx
+
+	// Get current latest block from blockchain.
+	latestBlockNum, err := t.src.GetLatestBlockNumber(ctx)
+	if err != nil {
+		return fmt.Errorf("getting latest block on %s: %w", t.state.NetworkName, err)
+	}
+
+	// Check if we already have the latest block.
+	if t.state.BlockNumber == latestBlockNum {
+		return ErrNothingNew
+	}
+
+	// Calculate how many blocks to retrieve.
+	delta := uint64(1)
+	if t.batchSize > 1 && latestBlockNum > t.state.BlockNumber {
+		delta = min(latestBlockNum-t.state.BlockNumber, t.batchSize)
+	}
+
+	// Determine range of blocks to process
+	startBlock := t.state.BlockNumber + 1
+	endBlock := startBlock + delta - 1
+
+	// TODO: This tx would update the block tracker to the previous block.
+	tx, err := t.db.Begin(ctx)
+	if err != nil {
+		return fmt.Errorf("beginning transaction: %w", err)
+	}
+	defer func() {
+		if err := tx.Rollback(); err != nil {
```

Comment:
technically you can call rollback on a committed transaction and it should not implode. But it might be cleaner if you check that a rollback should be called (for example err is not null)